<a href="https://colab.research.google.com/github/wiiiiish/2022-machine_learning/blob/main/ex02_%EC%A3%BC%ED%83%9D%EA%B0%80%EA%B2%A9_%EB%8D%B0%EC%9D%B4%ED%84%B0_%EC%98%88%EC%B8%A1%ED%95%98%EA%B8%B0.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### 목표
- sklearn 을 이용해 회귀 데이터를 학습해보자.
- LinearRegression 모델을 사용해보자.





### 1. 모델 정의

In [ ]:
from sklearn.linear_model import LinearRegression # 회귀 모델

In [ ]:
house_model = LinearRegression()

### 2. 모델 학습

##### 1. 데이터 로딩 

In [ ]:
import pandas as pd

In [ ]:
!pwd

/content


In [ ]:
%cd MyDrive/Colab\ Notebooks/2022-04-11\ 머신러닝

/content/drive/MyDrive/Colab Notebooks/2022-04-11 머신러닝


In [ ]:
# data폴더로 이동하고 melb_data.csv 파일을 pandas로 열어 상위 5개 살펴보자
house_data = pd.read_csv('./data/melb_data.csv', encoding='euc-kr')
house_data.head()

Suburb           Address  Rooms Type      Price Method SellerG  \
0  Abbotsford      85 Turner St      2    h  1480000.0      S  Biggin   
1  Abbotsford   25 Bloomburg St      2    h  1035000.0      S  Biggin   
2  Abbotsford      5 Charles St      3    h  1465000.0     SP  Biggin   
3  Abbotsford  40 Federation La      3    h   850000.0     PI  Biggin   
4  Abbotsford       55a Park St      4    h  1600000.0     VB  Nelson   

        Date  Distance  Postcode  ...  Bathroom  Car  Landsize  BuildingArea  \
0  3/12/2016       2.5    3067.0  ...       1.0  1.0     202.0           NaN   
1  4/02/2016       2.5    3067.0  ...       1.0  0.0     156.0          79.0   
2  4/03/2017       2.5    3067.0  ...       2.0  0.0     134.0         150.0   
3  4/03/2017       2.5    3067.0  ...       2.0  1.0      94.0           NaN   
4  4/06/2016       2.5    3067.0  ...       1.0  2.0     120.0         142.0   

   YearBuilt  CouncilArea Lattitude  Longtitude             Regionname  \
0        NaN        Yarra  -37.7996    144.9984  Northern Metropolitan   
1     1900.0        Yarra  -37.8079    144.9934  Northern Metropolitan   
2     1900.0        Yarra  -37.8093    144.9944  Northern Metropolitan   
3        NaN        Yarra  -37.7969    144.9969  Northern Metropolitan   
4     2014.0        Yarra  -37.8072    144.9941  Northern Metropolitan   

  Propertycount  
0        4019.0  
1        4019.0  
2        4019.0  
3        4019.0  
4        4019.0  

[5 rows x 21 columns]

In [ ]:
house_data.info() # 데이터 개수, 결측치 유무, 컬럼이름, 데이터타입

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13580 entries, 0 to 13579
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Suburb         13580 non-null  object 
 1   Address        13580 non-null  object 
 2   Rooms          13580 non-null  int64  
 3   Type           13580 non-null  object 
 4   Price          13580 non-null  float64
 5   Method         13580 non-null  object 
 6   SellerG        13580 non-null  object 
 7   Date           13580 non-null  object 
 8   Distance       13580 non-null  float64
 9   Postcode       13580 non-null  float64
 10  Bedroom2       13580 non-null  float64
 11  Bathroom       13580 non-null  float64
 12  Car            13518 non-null  float64
 13  Landsize       13580 non-null  float64
 14  BuildingArea   7130 non-null   float64
 15  YearBuilt      8205 non-null   float64
 16  CouncilArea    12211 non-null  object 
 17  Lattitude      13580 non-null  float64
 18  Longti

In [ ]:
# 결측치가 없고, 숫자로 구성된 데이터 중에 일부를 선택해 사용해보자.
X = house_data[['Rooms', 'Distance', 'Bathroom', 'Landsize', 'Lattitude', 'Longtitude']]
y = house_data['Price']

In [ ]:
X.shape, y.shape

((13580, 6), (13580,))

In [ ]:
# 훈련용과 평가용으로 데이터를 분리해보자. 비율은 8:2, randomstate = 412
from sklearn.model_selection import train_test_split # 훈련용, 평가용으로 데이터 분리

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=412)
X_train.head()

Rooms  Distance  Bathroom  Landsize  Lattitude  Longtitude
3022       4      13.7       1.0     692.0  -37.93860   145.00550
334        2       5.9       2.0       0.0  -37.77860   144.91410
9832       3      31.2       1.0     179.0  -38.07483   145.13133
10863      3      14.7       2.0     847.0  -37.88421   144.67869
10235      3      14.3       2.0     195.0  -37.79979   145.15665

In [ ]:
house_model.fit(X_train, y_train) # 지도학습 방법

LinearRegression()

In [ ]:
house_model.coef_ #선형 모델이 입력 특성에 대해 학습한 가중치

array([ 3.19904084e+05, -4.24902739e+04,  1.86938191e+05,  4.04283314e+00,
       -1.63108473e+06,  8.87040171e+05])

In [ ]:
house_model.intercept_ # 절편

-190009122.20245534

y = 319904.084 * Rooms - 42490.2739 * Distance + 186938.191 * Bathroom + 4.04283314 * Landsize - 1631084.73 * Lattitude + 887040.171 * Longtitude -190009122.20245534

### 3. 모델 예측

In [ ]:
pre = house_model.predict(X_test)

### 4. 모델 평가
- 집 가격은 데이터의 형태가 수치형이기 때문에 정답인지 아닌지 명확하지 않다. (개인 기준에 따라 달라질 수 있다.)
- 수치형 데이터는 예측값과 실제값의 차이인 오차를 이용해 모델을 평가한다.

In [ ]:
from sklearn.metrics import mean_absolute_error # 평균 절대값 오차

In [ ]:
house_mae = mean_absolute_error(y_test, pre) # 실제값, 예측값
house_mae

321470.49957908835

##### 다른 평가 지표

In [ ]:
house_model.score(X_test, y_test) # 평가용 문제, 평가용 답

0.44383536539265855

R^2 = 1 - 시그마(y-y_hat)^2 / 시그마(y-y_bar)^2